picard/log.py

```diff
@@ -176,6 +177,12 @@ def name_filter(record):
             path = path.resolve().relative_to(picard_module_path)
         except ValueError:
             pass
+        else:
+            if not DebugOpt.PLUGIN_FULLPATH.enabled and 'plugins' in path.parts:
+                parts = list(reversed(path.parts))
+                parts = parts[:parts.index('plugins') + 1]
+                path = Path(*reversed(parts))
+
```

Comment:
Perhaps I'm missing something, but won't this only shorten the paths if the path is already relative to the `picard` module (which will not be the case for plugins)?  I'm wondering if Lines 179-180 should be removed to shorten the paths that are not relative to the `picard` module?

Reply:
Yes, it only shortens file paths relative to picard module, for now.

We can extend it to apply to all paths, but it would be more "risky", imagine someone having picard installed this way:
`/some/path/plugins/picard/module/file.py` it would lead to  `plugins/picard/module`.

Instead we should perhaps use actual plugin paths, instead of simply matching on `plugins` in the path.

Reply:
I really must be missing something because I don't understand the issue.

> Yes, it only shortens file paths relative to picard module, for now.

So it only (potentially) shortens paths that have already been shortened because they are relative to the picard module?  Since the installed plugins are in a directory somewhere under the user's home directory (even when installed manually) -- with the possible exception of the Windows portable version, then it is **extremely** unlikely that they would ever be marked as relative to the picard module and thus would never be shortened.  In that case this new option to display the full path for plugins would be irrelevant because the full path to the plugins would always be displayed by default.

> We can extend it to apply to all paths, but it would be more "risky", imagine someone having picard installed this way: `/some/path/plugins/picard/module/file.py` it would lead to `plugins/picard/module`.

That's why I suggested that the new plugin shortening code would only be appled when the path was **not** relative to the picard module (when the exception was triggered) and the new `PLUGIN_FULLPATH.enabled` was false.

I don't understand your comment about it being more "risky".  I don't believe that this `name_filter()` function is used anywhere other than in the logger.  If your concern is that the displayed path to the plugin might be somehow confusing because it has been shortened, then isn't that why we're adding the option to display the full path?

> Instead we should perhaps use actual plugin paths, instead of simply matching on `plugins` in the path.

Are you saying that we should be displaying the full path to the plugins by default?  In that case, there would be no need to have this new option.

---

Perhaps a more acceptable solution would be to have the function first check if the path is relative to the picard module and only display starting at that directory (as in the original code).  Then if it is not relative to the picard module, check if it is relative to the current plugins directory (`USER_PLUGIN_DIR` from `picard.const`) and then display starting from that directory unless `PLUGIN_FULLPATH.enabled` was true.

Something like (untested and unoptimized code):

```python
from picard.const import USER_PLUGIN_DIR


def name_filter(record):
    # In case the module exists within picard, remove the picard prefix
    # else, in case of something like a plugin, keep the path as it is.
    # It provides a significant but short name from the filepath of the module
    path = Path(record.pathname).with_suffix('')
    # PyInstaller paths are already relative
    # FIXME: With Python 3.9 this should better use
    # path.is_relative_to(picard_module_path.parent)
    # to avoid the exception handling.
    if path.is_absolute():
        try:
            path = path.resolve().relative_to(picard_module_path)
        except ValueError:
            pass

        try:
            path = path.resolve().relative_to(USER_PLUGIN_DIR)
            if not DebugOpt.PLUGIN_FULLPATH.enabled and 'plugins' in path.parts:
                parts = list(reversed(path.parts))
                parts = parts[:parts.index('plugins') + 1]
                path = Path(*reversed(parts))
        except ValueError:
            pass

    parts = list(path.parts)
    if parts[-1] == '__init__':
        del parts[-1]
    if parts[0] == path.anchor:
        del parts[0]
    record.name = str(PurePosixPath(*parts))
    return True
```

Reply:
Note that I cleaned up the code a bit in my previous comment so that the logic makes a bit more sense.

Reply:
@rdswift I agree, I'll merge this, then you can submit a PR implementing it